client/src/erlang/receiver.zig

```diff
@@ -0,0 +1,311 @@
+                item.sentinel.?,
+            )
+        else
+            try allocator.alloc(
+                item.child,
+                usize_size,
+            );
+        errdefer allocator.free(slice_buffer);
+        for (slice_buffer) |*elem| {
+            elem.* = try internal_receive_message(item.child, allocator, deserializer);
+        }
+        try erl.validate(
+            error.decoding_list_in_pointer_2,
+            ei.ei_decode_list_header(deserializer.buf.buff, deserializer.index, &size),
+        );
+        if (size != 0) return error.decoded_improper_list;
+        value = slice_buffer;
+    }
+
+    return value;
+}
+
+inline fn receive_array(comptime T: type, comptime item: std.builtin.Type.Array, allocator: std.mem.Allocator, deserializer: Deserializer) !T {
+    var value: T = undefined;
+    var size: i32 = 0;
+    try erl.validate(
+        error.decoding_list_in_array_1,
+        ei.ei_decode_list_header(deserializer.buf.buff, deserializer.index, &size),
+    );
+    if (item.len != size) return error.wrong_array_size;
+    for (0..value.len) |idx| {
+        value[idx] = try internal_receive_message(item.child, allocator, deserializer);
+    }
+    try erl.validate(
+        error.decoding_list_in_array_2,
+        ei.ei_decode_list_header(deserializer.buf.buff, deserializer.index, &size),
+    );
+    if (size != 0) return error.decoded_improper_list;
+    return value;
+}
+
+inline fn receive_bool(comptime T: type, deserializer: Deserializer) !T {
+    var value: T = undefined;
+    var bool_value: i32 = 0;
+    try erl.validate(
+        error.decoding_boolean,
+        ei.ei_decode_boolean(deserializer.buf.buff, deserializer.index, &bool_value),
+    );
+    if (bool_value == 0) {
+        value = false;
+    } else {
+        value = true;
+    }
+    return value;
+}
+
+fn internal_receive_message(comptime T: type, allocator: std.mem.Allocator, deserializer: Deserializer) !T {
+    var value: T = undefined;
+    if (T == [:0]const u8) {
+        value = try receive_string(deserializer, allocator);
+    } else if (T == ei.erlang_pid) {
+        try erl.validate(
+            error.invalid_pid,
+            ei.ei_decode_pid(deserializer.buf.buff, deserializer.index, &value),
+        );
+    } else switch (@typeInfo(T)) {
+        .Struct => |item| {
+            value = try receive_struct(T, item, allocator, deserializer);
+        },
+        .Int => |item| {
+            value = try receive_int(T, item, deserializer);
+        },
+        .Enum => |item| {
+            value = try receive_enum(T, item, allocator, deserializer);
+        },
+        .Union => |item| {
+            value = try receive_union(T, item, allocator, deserializer);
+        },
+        .Pointer => |item| {
+            value = try receive_pointer(T, item, allocator, deserializer);
+        },
+        .Array => |item| {
+            value = try receive_array(T, item, allocator, deserializer);
+        },
```

Comment:
In general, I don't think we need to pass the `item` here into the nested function because it's just `@typeInfo(T)` unwrapped for a particular union tag.

Reply:
I checked all the internal functions and all of them use `item` aside from `@typeInfo(T)`. Are you sure?